arraykit.c

```diff
@@ -129,6 +130,62 @@ AK_ResolveDTypeIter(PyObject *dtypes)
     return resolved;
 }
 
+// Numpy implementation: https://github.com/numpy/numpy/blob/a14c41264855e44ebd6187d7541b5b8d59bb32cb/numpy/core/src/multiarray/methods.c#L1557
+PyObject*
+AK_ArrayCopy(PyObject *array, PyObject *memo)
+{
+    PyObject *id = PyLong_FromVoidPtr(array);
+    PyObject *found = PyDict_GetItem(memo, id);
+    if (found) { // found will be NULL if not in dict
+        Py_DECREF(id);
+        Py_INCREF(found); // got a borrowed ref
```

Comment:
So it's not an issue here, but I want to point out something. If `id` were an object tracked by the garbage collector (not an `int`, like we have here), this would be unsafe. Crazy, right?

The reason is that `Py_DECREF(id)` could cause arbitrary code to run (like a `__del__` method) that could do crazy things like removing `found` from `memo`, which would decrement it's refcount, and possibly even cause it to be cleaned up. Then `Py_INCREF(found)` blows up with a segfault.

It's maddening to try to wrap your mind around these things, but a general rule of thumb is:
- `Py_INCREF` is _always_ safe.
- `Py_DECREF` (or any C API that might use it, directly or indirectly (aka the whole C API)) might invoke arbitrary code that (amongst other things) makes any borrowed references junk. 

The fix would be simple: increment the borrowed reference _before_ calling `Py_DECREF`, by swapping these two lines.

Reply:
Thanks for this explanation!